src/protocol/boolean/solved_bits.rs

```diff
@@ -144,109 +144,82 @@ impl AsRef<str> for Step {
+                assert!(result.iter().all(Option::is_none));
+                continue; // without incrementing successes
+            }
 
-        // Reconstruct b_P
-        let b_p = (&s0.b_p, &s1.b_p, &s2.b_p).reconstruct();
+            let [s0, s1, s2] = result.map(Option::unwrap);
 
-        Ok(Some((b_b, b_p)))
-    }
+            // [b]_B must be the same bit lengths
+            assert_eq!(s0.b_b.len(), s1.b_b.len());
+            assert_eq!(s1.b_b.len(), s2.b_b.len());
 
-    #[tokio::test]
-    pub async fn fp31() -> Result<(), Error> {
-        let world = TestWorld::new().await;
-        let ctx = world.contexts::<Fp31>();
-        let [c0, c1, c2] = ctx;
+            // Reconstruct b_B from ([b_1]_p,...,[b_l]_p) bitwise sharings in F_p
+            let b_b = (0..s0.b_b.len())
+                .map(|i| {
+                    let bit = (&s0.b_b[i], &s1.b_b[i], &s2.b_b[i]).reconstruct();
+                    assert!(bit == F::ZERO || bit == F::ONE);
+                    bit
+                })
+                .collect::<Vec<_>>();
 
-        let mut success = 0;
-        for i in 0..21 {
-            let record_id = RecordId::from(i);
-            if let Some((b_b, b_p)) =
-                random_bits([c0.clone(), c1.clone(), c2.clone()], record_id).await?
-            {
-                // Base10 of `b_B ⊆ Z` must equal `b_P`
-                assert_eq!(b_p.as_u128(), bits_to_value(&b_b));
-                success += 1;
-            }
+            // Reconstruct b_P
+            let b_p = (&s0.b_p, &s1.b_p, &s2.b_p).reconstruct();
+
+            // Base10 of `b_B ⊆ Z` must equal `b_P`
+            assert_eq!(b_p.as_u128(), bits_to_value(&b_b));
+
+            successes += 1;
         }
-        // The chance of this protocol aborting 21 out of 21 tries in Fp31
-        // is about 2^-100. Assert that at least one run has succeeded.
-        assert!(success > 0);
 
-        Ok(())
+        assert!(successes > 0);
     }
 
     #[tokio::test]
-    pub async fn fp_32bit_prime() -> Result<(), Error> {
-        let world = TestWorld::new().await;
-        let ctx = world.contexts::<Fp32BitPrime>();
-        let [c0, c1, c2] = ctx;
-
-        let mut success = 0;
-        for i in 0..4 {
-            let record_id = RecordId::from(i);
-            if let Some((b_b, b_p)) =
-                random_bits([c0.clone(), c1.clone(), c2.clone()], record_id).await?
-            {
-                // Base10 of `b_B ⊆ Z` must equal `b_P`
-                assert_eq!(b_p.as_u128(), bits_to_value(&b_b));
-                success += 1;
-            }
-        }
-        // The chance of this protocol aborting 4 out of 4 tries in Fp32BitPrime
-        // is about 2^-100. Assert that at least one run has succeeded.
-        assert!(success > 0);
+    pub async fn fp31() {
+        random_bits::<Fp31, 21>().await;
+    }
 
-        Ok(())
+    #[tokio::test]
+    pub async fn fp_32bit_prime() {
+        random_bits::<Fp32BitPrime, 4>().await;
```

Comment:
Maybe comment that:

```rust
// Probability of failure for one iteration is 5/2^32 =~ 2^-30.
// Need 4 runs to reduce failure to < 2^-100.

Reply:
If there's an effort to avoid unnecessary loops (to keep the test suite fast?), then maybe it makes sense to change these to run up to the specified number of iterations, stopping on the first success?

(This potentially increases test non-determinism, but I don't think in a way that matters. There is a similar case where a RandomBitsGenerator is created with a total record count -- that will fail intermittently with record_id out of range when there is a solved_bits failure. I added an assertion in `RandomBitsGenerator::new` to catch this. But the truly equivalent thing here would be doing random_bits exactly once and expecting it to succeed -- then the test would intermittently fail. I can't think of a problem with stopping on the first success.)